### DISCUSSION TITLE
Opening shell in current DirSheet

### DISCUSSION THREAD

**Discussion by cool-RR** [saulpw/visidata#2086] [Q&A] [answered]
I have a `DirSheet` open. I want to be dropped into a shell (Bash or zsh or whatever other shell I'm using) with that folder as the active folder. I want to run as many commands as I want, and when I press Ctrl-D, I want to return to VisiData. Does something like that exists?

**Comment by saulpw** [comment, marked as answer]
```
DirSheet.addCommand('', 'sysopen-shell', '''
import subprocess
with SuspendCurses():
    subprocess.call(os.getenv('SHELL'), cwd=str(source))
''')
```

**Reply by cool-RR** [reply]
Woohoo! That works perfectly.

**Comment by saulbert** [comment]
VisiData's `develop` branch now has an `exec-shell` command, shipping in v3.4. On any sheet it suspends the display, launches `$SHELL`, and returns on `Ctrl+D` — the flow described here. Invoke it by longname from the command palette, or from the **System › Open shell** menu (no default keybinding).

One difference from the original request: `exec-shell` starts the shell in VisiData's own working directory, not the current sheet's source folder. For the DirSheet-in-its-own-folder behavior, the `cwd=str(source)` snippet above still does that; a per-sheet working directory for `exec-shell` could be a follow-up.

[this message written by Claude Opus 4.8 and approved by @saulpw]

**Reply by cool-RR** [reply]
Great, thank you!
